cl/recap/tasks.py

```diff
@@ -1172,15 +1186,24 @@ def fetch_pacer_doc_by_rd(
         return
 
     pacer_case_id = rd.docket_entry.docket.pacer_case_id
-    try:
-        r, r_msg = download_pacer_pdf_by_rd(
-            rd.pk, pacer_case_id, rd.pacer_doc_id, cookies, magic_number
-        )
-    except (requests.RequestException, HTTPError):
-        msg = "Failed to get PDF from network."
-        mark_fq_status(fq, msg, PROCESSING_STATUS.FAILED)
-        self.request.chain = None
-        return
+
+    # If a response containing the PDF data is not provided, try to download
+    # the PDF document by rd.
+    if not r:
+        try:
+            r, r_msg = download_pacer_pdf_by_rd(
+                rd.pk,
+                pacer_case_id,
+                rd.pacer_doc_id,
+                cookies,
+                magic_number,
+                appellate,
+            )
+        except (requests.RequestException, HTTPError):
+            msg = "Failed to get PDF from network."
+            mark_fq_status(fq, msg, PROCESSING_STATUS.FAILED)
+            self.request.chain = None
+            return
```

Comment:
I'm sorry, I find this kind of ugly. The main problem is that by putting this section in an `if` block, you're changing the whole meaning of the function name (it's no longer fetching!), but beyond that when this section isn't called because you already have the response object, you don't really need this function. I can see how this is sort of simple. It's really just a one-line fix to accomplish what we need (just add the `if` statement), but it makes the code more confusing. 

I think it's much better to branch the code earlier, in `process_recap_email`, where you do your `if appellate` branch. At that point, we should just have composable pieces that we can work with:

```
if appellate:
    get_document_number_appellate(...)
else:
    fetch_pacer_doc_by_rd(...)
```

I think if you do it that way, you can add any logic that's currently in `fetch_pacer_doc_by_rd` into the `get_document_number_appellate` — and you can rename `get_document_number_appellate` to be `fetch_pacer_doc_appellate` or something like that. I think it should vastly simplify this section.

Reply:
Yeah, that's true. I also didn't like to much that tweak to `fetch_pacer_doc_by_rd` I wasn't sure if in this case, it was better than repeating some code that's within `fetch_pacer_doc_by_rd`. 

Thanks for the advice, so yes I think here is better to simplify things. So now I've renamed `get_document_number_appellate` to `fetch_pacer_doc_appellate` and created a new method  `save_pacer_doc_appellate ` so now:

`fetch_pacer_doc_appellate`: Tries to download the PDF document to get the document number from it or from the download confirmation page in case is not possible to get it from the PDF.

`save_pacer_doc_appellate`: Since we first need the document number in order to create the docket entry/recap document and save the PDF document. I think it makes sense to have a different function to save PDF file once the docket entry/recap document are ready. So `save_pacer_doc_appellate` save the file from the HTTP response containing the PDF returned by `fetch_pacer_doc_appellate`. Now things are separated and simpler.